src/ripple/app/paths/impl/StrandFlow.h

```diff
@@ -817,6 +817,35 @@ flow(
     JLOG(j.trace()) << "Total flow: in: " << to_string(actualIn)
                     << " out: " << to_string(actualOut);
 
+    /* flowCross doesn't handle offer crossing with tfFillOrKill flag correctly.
+     * 1. If tfFillOrKill is set then the owner must receive the full
+     *   TakerPays. We reverse pays and gets because during crossing
+     *   we are taking, therefore the owner must deliver the full TakerPays and
+     *   the entire TakerGets doesn't have to be spent.
+     *   Pre-fixFillOrKill amendment code fails if the entire TakerGets
+     *   is not spent. fixFillOrKill addresses this issue.
+     * 2. If tfSell is also set then the owner must spend the entire TakerGets
+     *   even if it means obtaining more than TakerPays. Since the pays and gets
+     *   are reversed, the owner must send the entire TakerGets.
+     */
+    bool const fixFillOrKill =
+        baseView.rules().enabled(fixFillOrKillOnFlowCross);
+    // tfSell is handled by setting the deliver amount to max
+    auto const sell = [&]() -> bool {
+        if constexpr (std::is_same_v<TOutAmt, XRPAmount>)
+        {
+            static auto max = XRPAmount{STAmount::cMaxNative};
+            return outReq == max;
```

Comment:
We can talk about this, but I don't love this solution. It's using a special value of output request to change behavior. Any payment that uses this value will get this special behavior, even payments that aren't offers and don't have the flag set. It also means if the implementation changes so we no long set these max amounts this code breaks.

Reply:
I can't argue that this is a great solution, but that's a problem with the FlowCross implementation, not with this fix.  @gregtatcam is using the only indication that FlowCross sends down through the payment engine that the `tfSell` flag is set:

https://github.com/XRPLF/rippled/blob/develop/src/ripple/app/tx/impl/CreateOffer.cpp#L739-L755

But, yes, there is also an `offerCrossing` flag that is passed as well.  We could add that to the `sell` detection.  Maybe like this...
```
    // FlowCross handles tfSell by setting the deliver amount to max
    auto const sell = [&]() -> bool {
        if (offerCrossing)
        {
            if constexpr (std::is_same_v<TOutAmt, XRPAmount>)
            {
                static const auto max = XRPAmount{STAmount::cMaxNative};
                return outReq == max;
            }
            else if constexpr (std::is_same_v<TOutAmt, IOUAmount>)
            {
                static const auto max =
                    IOUAmount{STAmount::cMaxValue / 2, STAmount::cMaxOffset};
                return outReq >= max;
            }
        }
        return false;
    }();
```
That's still not great, but it makes sure that the `sell` flag is only set when offer crossing.

Reply:
Thanks @scottschurr, it is better but as you said not great. I wanted to minimize the payment engine impact and not change the arguments to `flow()`. If it's ok to change the args then the `offerCrossing` can be changed from `bool` to `optional<uint32>`, which if seated contains the offer crossing flags and if not seated then it's not offer crossing. The semantics of the `offerCrossing` flag remains the same but it does provide more information. The only thing I don't like is that `partialPayment` flag carries a duplicate info (it's set to `!(txFlags & tfFillOrKill)`) but it's probably ok. @seelabs does this work?

Reply:
Looks good. Thanks for making that change, I appreciate it.